int32 right_value
int32 left_value
int32 up_value
int32 down_value
